core/signer/signer.go

```diff
@@ -94,110 +71,129 @@ func (s *Signer) StartSigning(ctx context.Context, interval time.Duration) {
+		if leaf, ok := leafMap[toArray(index)]; ok {
+			oldValue = leaf.LeafValue
+		}
+
+		// TODO: change mutator interface to accept objects directly.
+		mData, err := proto.Marshal(m)
 		if err != nil {
-			return 0, err
+			return nil, err
 		}
-		if err := s.queueMutation(txn, mutation.GetKeyValue().Key, mData); err != nil {
-			return 0, fmt.Errorf("queueMutation err: %v", err)
+		newValue, err := s.mutator.Mutate(oldValue, mData)
+		if err != nil {
+			log.Printf("Mutate(): %v", err)
+			continue // A bad mutation should not make the whole batch fail.
+		}
+
+		retMap[toArray(index)] = &trillian.MapLeaf{
+			Index:     index,
+			LeafValue: newValue,
 		}
 	}
-	return int64(maxSequence), nil
+	// Convert return map back into a list.
+	ret := make([]*trillian.MapLeaf, 0, len(retMap))
+	for _, v := range retMap {
+		ret = append(ret, v)
+	}
+	return ret, nil
 }
 
 // CreateEpoch signs the current map head.
 func (s *Signer) CreateEpoch(ctx context.Context) error {
-	txn, err := s.factory.NewTxn(ctx)
+	// Get the current root.
+	startSequence := int64(0)
+	rootResp, err := s.tmap.GetSignedMapRoot(ctx, &trillian.GetSignedMapRootRequest{
+		MapId: s.mapID,
+	})
 	if err != nil {
-		return fmt.Errorf("NewTxn(): %v", err)
-	}
-	if err := s.createEpoch(ctx, txn); err != nil {
-		if err := txn.Rollback(); err != nil {
-			log.Printf("Cannot rollback the transaction: %v", err)
-		}
-		return fmt.Errorf("createEpoch(): %v", err)
-	}
-	if err := txn.Commit(); err != nil {
-		return fmt.Errorf("txn.Commit(): %v", err)
+		return fmt.Errorf("GetSignedMapRoot(%v): %v", s.mapID, err)
 	}
-	return nil
-}
+	startSequence = rootResp.GetMapRoot().GetMetadata().GetHighestFullyCompletedSeq()
 
-func (s *Signer) createEpoch(ctx context.Context, txn transaction.Txn) error {
-	maxSequence, err := s.processMutations(ctx, txn)
+	// Get the list of new mutations to process.
+	mutations, seq, err := s.newMutations(ctx, startSequence)
 	if err != nil {
-		return fmt.Errorf("processMutations(): %v", err)
+		return fmt.Errorf("newMutations(%v): %v", startSequence, err)
 	}
-	if err := s.tree.Commit(txn); err != nil {
-		return fmt.Errorf("Commit(): %v", err)
-	}
-	epoch, err := s.tree.Epoch(txn)
-	if err != nil {
-		return fmt.Errorf("Epoch(): %v", err)
+
+	// Get current leaf values.
+	indexes := make([][]byte, 0, len(mutations))
+	for _, m := range mutations {
+		indexes = append(indexes, m.KeyValue.Key)
 	}
-	root, err := s.tree.ReadRootAt(txn, epoch)
+	getResp, err := s.tmap.GetLeaves(ctx, &trillian.GetMapLeavesRequest{
+		MapId: s.mapID,
+		Index: indexes,
+	})
 	if err != nil {
-		return fmt.Errorf("ReadRootAt(%v): %v", epoch, err)
+		return err
 	}
 
-	smr := &trillian.SignedMapRoot{
-		MapId:          0,
-		TimestampNanos: s.clock.Now().UnixNano(),
-		RootHash:       root,
-		MapRevision:    epoch,
-		Metadata: &trillian.MapperMetadata{
-			HighestFullyCompletedSeq: maxSequence,
-		},
+	// TODO: verify inclusion proofs?
```

Comment:
I'm not sure we need to verify inclusion proofs here, hence the question mark. 
If the map server was being run by a different entity, we'd want to verify the inclusion proofs.  However, in our case, we expect the keyserver and the mapserver to be run by the same entity.